test/unit/bitmap_test.c

```diff
@@ -0,0 +1,244 @@
+            return false;
+        }
+    }
+    deallocate_bitmap(b_cpy);
+    return true;
+}
+
+/**
+ *  Tests copying of bitmap using bitmap_copy
+ *  function.
+ *  Also implicity tests bitmap_foreach_set
+ */ 
+boolean test_copy(heap h, bitmap b) { 
+    bitmap b_cpy = allocate_bitmap(h, h, 4096);
+    for (int i = 0; i < 20; i++) {
+        bitmap_set(b_cpy, rand(), 1);
+        bitmap_set(b, rand(), 1);
+    }
+    bitmap_copy(b, b_cpy);
+    // tests bits match in original bitmap
+    bitmap_foreach_set(b, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if (!bitmap_get(b_cpy, j)) {
+            msg_err("!!! cloning failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    // tests bits match in copy of bitmap
+    bitmap_foreach_set(b_cpy, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b_cpy, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! cloning failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    deallocate_bitmap(b_cpy);
+    return true;
+}
+
+/**
+ *  Tests getting and setting bits of a bitmap using 
+ *  bitmap_get and bitmap_set functions.
+ */ 
+boolean test_set_and_get(bitmap b) {
+    u64 i = rand();
+    bitmap_set(b, i, 1);
+    if (!bitmap_get(b, i)) {
+        msg_err("!!! set and get failed for bitmap\n");
+        return false;
+    }
+    return true;
+}
+
+/**
+ *  Tests wrapping and unwrapping of bitmap using 
+ *  bitmap_wrap and bitmap_unwrap functions.
+ */ 
+boolean test_wrap(heap h) { 
+    u64 map = rand(); 
+    map = (map << 32) | rand();
+    bitmap b = bitmap_wrap(h, &map, 64);
+    for (int i = 0; i < 64; i++) {
+        if (((map & (1ULL << i)) && !bitmap_get(b, i)) || 
+            (!(map & (1ULL << i)) && bitmap_get(b, i))) {
+            msg_err("!!! wrap failed for bitmap at bit %d | map: %ld bitmap: %ld\n", 
+                i, (map & (1ULL << i)), bitmap_get(b, i));
+            bitmap_unwrap(b);
+            return false;
+        }
+    }
+    bitmap_unwrap(b);
+    return true;
+}
+
+/**
+ *  Tests bitmap alloc and dealloc of bitmap using 
+ *  bitmap_alloc_within_range and bitmap_dealloc functions.
+ */ 
+boolean test_bitmap_alloc(bitmap b, u64 start, u64 end) {
+    u64 nbits = rand();
+    u64 first;
+    while (true) {
+        first = bitmap_alloc_within_range(b, nbits, start, end);
+        if (first != INVALID_PHYSICAL) {
+            if (nbits > (end-start))
+                msg_err("!!! invalid value for nbits: %ld end-start: %ld\n", nbits, (end-start));
+            break;
```

Comment:
If (nbits > end - start), this function should return false, so that the test fails